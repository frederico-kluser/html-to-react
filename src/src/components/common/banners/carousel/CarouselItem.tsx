/* eslint-disable @next/next/no-img-element */
import { IMedia } from '@/types/content';

import React from 'react';

const renderCrouselImageItem = (media: IMedia) => {
  const { imageUrl, href } = media;

  const imageSrc = imageUrl ?? '';
  const image = (
    <img data-testid={`carousel-image`} src={imageSrc} alt={`banner`} />
  );

  if (!href) {
    return image;
  }

  return (
    <a data-testid={`carousel-link`} href={href}>
      {image}
    </a>
  );
};

export const CarouselItem = (media: IMedia) => {
  if (media.imageUrl) {
    return renderCrouselImageItem(media);
  }
  return null;
};

export default CarouselItem;
