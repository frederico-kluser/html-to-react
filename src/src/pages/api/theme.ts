import { connectToDatabase } from '../../database/connect';
import type { NextApiRequest, NextApiResponse } from 'next';

export const getTheme = async () => {
  const { db } = await connectToDatabase();
  return await db.collection('themes').findOne({ name: 'Light' });
};

const handler = async (req: NextApiRequest, res: NextApiResponse) => {
  const theme = await getTheme();
  res.status(200).json(theme);
};

export default handler;
