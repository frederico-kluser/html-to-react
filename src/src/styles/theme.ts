export const theme = {
  id: 'T1',
  name: 'Light',
  layout: {
    type: 'fixed',
    maxWidth: 1200,
    widthUnit: 'px'
  },
  colors: {
    primary: {
      main: '#0C1446',
      common: {
        black: '#000',
        white: '#fff',
        gray: '#616161',
        bgColor: '#af9995'
      }
    }
  },
  breakpoints: {
    size: {
      xs: 0,
      sm: 600,
      md: 960,
      lg: 1280,
      xl: 1920
    },
    unit: 'px'
  }
};
