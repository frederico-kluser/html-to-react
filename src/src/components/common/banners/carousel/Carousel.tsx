/* eslint-disable @next/next/no-img-element */
import { default as MultiCarousel } from 'react-multi-carousel';
import { ContainerFixed, ContainerFluid } from './Carousel.styles';
import { responsiveConfig } from './Carousel.config';
import { CarouselItem } from './CarouselItem';
import { IBanner } from '@models/content';

export type CarouselProps = IBanner & {
  /**
   * renders observing the max width defined in the theme or
   * uses all available horizontal space - default 'fixed'
   */
  layoutType?: 'fixed' | 'fluid';
  /**
   * Enable auto play mode. Default true
   */
  autoPlay?: boolean;
  /**
   * Hide/Show the default arrows - default true
   */
  arrows?: boolean;
  /**
   * Auto play speed in ms. Default 3000
   */
  autoPlaySpeed?: number;
  /**
   * Shows the next items and previous items partially. Default: false
   */
  centerMode?: boolean;
  /**
   * Infinite loop
   */
  infinite?: boolean;
  /**
   * Show/Hide the default dot list
   */
  showDots?: boolean;
};

const Carousel = (props: CarouselProps) => {
  const { medias, layoutType = 'fixed' } = props;

  const {
    arrows = true,
    autoPlay = true,
    autoPlaySpeed = 3000,
    centerMode = false,
    infinite = true
  } = props;

  const Container = layoutType === 'fixed' ? ContainerFixed : ContainerFluid;

  const sortedMedias = medias.sort((a, b) => b.order - a.order);

  return (
    <Container data-testid={`carousel-banner-${layoutType}`}>
      <MultiCarousel
        arrows={arrows}
        autoPlay={autoPlay}
        autoPlaySpeed={autoPlaySpeed}
        centerMode={centerMode}
        draggable
        focusOnSelect={false}
        infinite={infinite}
        keyBoardControl
        minimumTouchDrag={80}
        renderButtonGroupOutside={false}
        renderDotsOutside={false}
        responsive={responsiveConfig}
        showDots={false}
        slidesToSlide={1}
        swipeable
        itemClass="carousel-item"
      >
        {sortedMedias.map((media, index) => {
          return <CarouselItem key={index} {...media} />;
        })}
      </MultiCarousel>
    </Container>
  );
};

export default Carousel;
