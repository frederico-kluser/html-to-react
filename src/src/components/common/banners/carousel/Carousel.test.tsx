import { render, screen, cleanup } from '@testing-library/react';
import { ReactNode } from 'react';
import { ThemeProvider } from 'styled-components';
import { theme } from '../../../../styles/theme';
import Carousel from '.';
import {
  bannerCarousel,
  bannerCarouselWithoutHref,
  bannerCarouselWithoutImage
} from './__mocks__/caroulselContent';
import CarouselItem from './CarouselItem';

afterEach(cleanup);

const renderComponent = (children: ReactNode) =>
  render(<ThemeProvider theme={theme}>{children}</ThemeProvider>);

describe('<Carousel />', () => {
  it('Should render the carousel with default layout fixed', () => {
    const { container } = renderComponent(<Carousel {...bannerCarousel} />);

    expect(screen.getByTestId(`carousel-banner-fixed`)).toBeInTheDocument();

    expect(container.firstChild).toMatchSnapshot();
  });

  it('Should render the corousel with layout fluid', () => {
    const { container } = renderComponent(
      <Carousel {...bannerCarousel} layoutType="fluid" />
    );

    expect(screen.getByTestId(`carousel-banner-fluid`)).toBeInTheDocument();

    expect(container.firstChild).toMatchSnapshot();
  });
});

describe('<CarouselItem />', () => {
  it('Should render a carousel Image Item with link', () => {
    const { medias } = bannerCarousel;
    const [media] = medias;

    const { container } = renderComponent(<CarouselItem {...media} />);

    expect(screen.getByTestId(`carousel-image`)).toBeInTheDocument();
    expect(screen.getByTestId(`carousel-link`)).toBeInTheDocument();

    expect(container.firstChild).toMatchSnapshot();
  });

  it('Should render a carousel Image Item without link', () => {
    const { medias } = bannerCarouselWithoutHref;
    const [media] = medias;

    const { container } = renderComponent(<CarouselItem {...media} />);

    expect(screen.getByTestId(`carousel-image`)).toBeInTheDocument();

    expect(container.firstChild).toMatchSnapshot();
  });

  it('Should handle an image without  source url', () => {
    const { medias } = bannerCarouselWithoutImage;
    const [media] = medias;

    const { container } = renderComponent(<CarouselItem {...media} />);

    expect(container.firstChild).toBeFalsy();

    expect(container.firstChild).toMatchSnapshot();
  });
});
