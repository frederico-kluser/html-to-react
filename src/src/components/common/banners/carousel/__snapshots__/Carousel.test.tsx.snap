// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Carousel /> Should render the carousel with default layout fixed 1`] = `
.c0 {
  margin: auto;
  max-width: 1200px;
}

.c0 img {
  margin: auto;
  max-width: 100%;
}

<div
  class="c0"
  data-testid="carousel-banner-fixed"
>
  <div
    class="react-multi-carousel-list  "
  >
    <ul
      class="react-multi-carousel-track "
      style="transition: none; overflow: unset; transform: translate3d(0px,0,0);"
    />
  </div>
</div>
`;

exports[`<Carousel /> Should render the corousel with layout fluid 1`] = `
.c0 {
  width: 100%;
}

.c0 .carousel-item {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c0 .carousel-item img {
  margin: auto;
  max-width: 100%;
}

<div
  class="c0"
  data-testid="carousel-banner-fluid"
>
  <div
    class="react-multi-carousel-list  "
  >
    <ul
      class="react-multi-carousel-track "
      style="transition: none; overflow: unset; transform: translate3d(0px,0,0);"
    />
  </div>
</div>
`;

exports[`<CarouselItem /> Should handle an image without  source url 1`] = `null`;

exports[`<CarouselItem /> Should render a carousel Image Item with link 1`] = `
<a
  data-testid="carousel-link"
  href="https://www.amazon.com/b/?_encoding=UTF8&node=23508887011&pf_rd_r=NGQQXRSZEX2BTSSXFMSE&pf_rd_p=b3367632-3703-47fc-9e4a-042ab9bd1208"
>
  <img
    alt="banner"
    data-testid="carousel-image"
    src="https://m.media-amazon.com/images/I/61CiqVTRBEL._SX1500_.jpg"
  />
</a>
`;

exports[`<CarouselItem /> Should render a carousel Image Item without link 1`] = `
<img
  alt="banner"
  data-testid="carousel-image"
  src="https://m.media-amazon.com/images/I/61CiqVTRBEL._SX1500_.jpg"
/>
`;
